src/cockpit/channels/filesystem.py

```diff
@@ -150,110 +158,75 @@ def do_yield_data(self, options: JsonObject) -> GeneratorChannel.DataGenerator:
-                self.unlink_temppath()
-                raise
-
-        self._tempfile.write(data)
-
-    def do_done(self):
-        if self._tempfile is None:
-            try:
-                os.unlink(self._path)
-            # crash on other errors, as they are unexpected
-            except FileNotFoundError:
-                pass
-        else:
-            self._tempfile.flush()
-
-            if self._tag and self._tag != tag_from_path(self._path):
-                raise ChannelError('change-conflict')
-
-            try:
-                os.rename(self._temppath, self._path)
-            # ensure to not leave the temp file behind
-            except FileNotFoundError as exc:
-                self.unlink_temppath()
-                raise ChannelError('not-found', message=str(exc)) from exc
-            except IsADirectoryError as exc:
-                self.unlink_temppath()
-                # not ideal, but the closest code we have
-                raise ChannelError('access-denied', message=str(exc)) from exc
-            except OSError as exc:
-                self.unlink_temppath()
-                raise ChannelError('internal-error', message=str(exc)) from exc
-
-            self._tempfile.close()
-            self._tempfile = None
+                # otherwise, spool data into a temporary file until EOF then rename into place...
+                with tempfile.NamedTemporaryFile(dir=dirname, prefix=f'.{basename}-', delete=False) as tmp:
+                    loop = asyncio.get_running_loop()
+                    while data is not None:
+                        await loop.run_in_executor(None, tmp.write, data)
+                        data = await self.read()
+
+                    os.fdatasync(tmp.fileno())
+
+                    if tag is None:
+                        # no preconditions about what currently exists or not
+                        # calculate the file mode from the umask
+                        os.fchmod(tmp.fileno(), 0o666 & ~my_umask())
+                        os.rename(tmp.name, path)
+                        tmp = None  # don't need to delete
+
+                    elif tag == '-':
+                        # the file must not exist.  file mode from umask.
+                        os.fchmod(tmp.fileno(), 0o666 & ~my_umask())
+                        os.link(tmp.name, path)  # will fail if file exists
+
+                    else:
+                        # the file must exist with the given tag
+                        buf = os.stat(path)
+                        if tag != tag_from_stat(buf):
+                            raise ChannelError('change-conflict')
+                        # chown/chmod from the existing file permissions
+                        os.fchmod(tmp.fileno(), stat.S_IMODE(buf.st_mode))
+                        os.fchown(tmp.fileno(), buf.st_uid, buf.st_gid)
+                        os.rename(tmp.name, path)
+                        tmp = None  # don't need to delete
```

Comment:
Wow, is that safe with a `with` statement?  This looks like it could seriously crash the contex's `__exit__()`. It apparently doesn't right now, but this is too hackish for my taste.

I just tested
```
>>> import tempfile, os
>>> with tempfile.NamedTemporaryFile() as tmp:
...     os.unlink(tmp.name)
```
and that works fine. There is no situation where we would *ever* want to keep the temporary file around, it should be removed in all cases. So it seems that can be cleaned up?

Reply:
Does that also work in Python 3.6?  At some point this gave `FileNotFoundError`.  They recently improved the situation with a new `delete_on_close` kwarg in Python 3.12, so I wouldn't be surprised if there were other cleanups on this point.

We could do the delete and suppress the message, but there's a theoretical (albeit effectively 0) possibility that we're deleting a file that we didn't create in that case:

 - we create the tempfile, write write, fsync
 - rename it into place over the real filename
 - at this exact moment another process comes along and creates a new temp file and just happens to get exactly the same name as the one we used
 - we `os.unlink()` what we think is our filename, but is actually the one created by the other process

So effectively, I'm using `tmp = None` here to mean "don't delete this".

As for the context manager vs. `None` thing, this isn't actually a problem.  There are two things involved in a context manager:

 - the context manager itself (which is captured by the `with` block to have `__exit__()` called later)
 - the return value of the `__enter__()` function of the context manager — this is the thing that gets optionally captured by the optional `as` clause

We're nulling out the second copy, but not the first.  In this case, I *think* it's the same object, but to be honest I'm not sure.  In either case, it doesn't matter — `with` holds its own copy and will `.close()` it (but not delete it).

I had a different version of this code that kept a separate variable which was the name of the file that we should delete when exiting the function.  I could try that again if you feel it is better.  I ended up liking this version more.

Reply:
Thanks for the explanation. Makes sense to keep an internal copy, so that the code can mess around with the variable. 

I tested it on py3.6 (rhel 8.1.0) and that indeed fails with a `FileNotFoundError`.

All that umask/weird scoping here feels like that NamedTemporaryFile isn't really the API that we want? Replacing one hack (tmpname calculation) with two others (smaller, but harder to understand) doesn't really feel like an improvement to me.

Just to gauge: I don't like it, but not a veto. I just wanted to understand how this works. If you like it, then go ahead, but I wouldn't mind keeping the current approach for this (the async conversion is really nice, though, but that's independent).

Reply:
So there's another consideration here which I didn't explicitly state that made me lean more towards `NamedTemporaryFile`.  I *really* dislike the chown/chmod tricks (as you know), and I particularly dislike the idea that we might be in a situation where either:
 - we chown/chmod late, in which case we might temporarily reveal sensitive data which ought not to be revealed (which is a security problem)
 - we chown/chmod early, in which case we might have some file with special bits like setuid in a half-written state (which is a security problem)

So if we're going to `chown/chmod`, then I specifically like that `NamedTemporaryFile` creates a file as the current user, with a maximally locked down set of permissions, and we only change the file mode/owner after it's fully written.  It actually legit seems like the best way.

But ya — the inability to deal with umask nicely from Python is a long-documented and awful problem.  There's a similar problem where it's very difficult to create executable files with the proper permissions because `open()` doesn't support the executable bit and you can't `chmod()` after the fact because of the frustration of determining the value of the umask.  The solution in that case is the same, and it sucks — lose all of the benefits of the standard library and write low-level code via `os.open()`.

Reply:
Note: we can do the "start with minimal permissions" fairly easily in the current code as well if we change it to
```
fd = os.open(..., 0o666 if stat is None else 0600)
```
and moving the chown to closing.